e2e_tests.py

```diff
@@ -93,95 +109,120 @@ def invalidMap(tag, method, path, bodies):
-    ['valid login after password change', 'post', '/auth/login', {}, {'username': username + '@domain.com', 'password': 'foobar2'}, 200, successfulLogin],
-    ['logout', 'post', '/auth/logout', {}, {}, 200],
-    ['check that session is no longer valid', 'get', '/profile', {}, '', 403],
-    ['login again', 'post', '/auth/login', {}, {'username': username, 'password': 'foobar2'}, 200, successfulLogin],
-    invalidMap ('change password', 'post', '/auth/change_password', ['', [], {}, {'foo': 'bar'}, {'old_password': 1}, {'old_password': 'foobar'}, {'old_password': 'foobar', 'new_password': 1}]),
-    ['get profile before profile update', 'get', '/profile', {}, {}, 200, getProfile1],
-    invalidMap ('update profile', 'post', '/profile', ['', [], {'email': 'foobar'}, {'birth_year': 'a'}, {'birth_year': 20}, {'country': 'Netherlands'}, {'gender': 0}, {'gender': 'a'}]),
-    ['change profile with same email', 'post', '/profile', {}, {'email': username + '@domain.com', 'country': 'US'}, 200],
-    ['change profile with different email', 'post', '/profile', {}, {'email': username + '@domain2.com', 'country': 'NL'}, 200, emailChange],
-    ['get profile after profile update', 'get', '/profile', {}, {}, 200, getProfile2],
-    ['verify email after email change', 'get', lambda state: '/auth/verify?' + urllib.parse.urlencode ({'username': username, 'token': state ['token2']}), {}, '', 302],
-    ['get profile after email verification', 'get', '/profile', {}, {}, 200, getProfile3],
-    invalidMap ('recover password', 'post', '/auth/recover', ['', [], {}, {'username': 1}]),
-    ['recover password, invalid user', 'post', '/auth/recover', {}, {'username': 'nosuch'}, 403],
-    ['recover password', 'post', '/auth/recover', {}, {'username': username}, 200, recoverPassword],
-    invalidMap ('reset password', 'post', '/auth/reset', ['', [], {}, {'username': 1}, {'username': 'foobar', 'token': 1}, {'username': 'foobar', 'token': 'some'}, {'username': 'foobar', 'token': 'some', 'password': 1}, {'username': 'foobar', 'token': 'some', 'password': 'short'}]),
-    ['reset password, invalid username', 'post', '/auth/reset', {}, lambda state: {'username': 'foobar', 'token': state ['token'], 'password': 'foobar'}, 403],
-    ['reset password, invalid token', 'post', '/auth/reset', {}, lambda state: {'username': username, 'token': 'foobar', 'password': 'foobar'}, 403],
-    ['reset password, invalid password', 'post', '/auth/reset', {}, lambda state: {'username': username, 'token': state ['token'], 'password': 'short'}, 400],
-    ['reset password', 'post', '/auth/reset', {}, lambda state: {'username': username, 'token': state ['token'], 'password': 'foobar3'}, 200],
-    ['login again after reset', 'post', '/auth/login', {}, {'username': username, 'password': 'foobar3'}, 200],
-    ['destroy account', 'post', '/auth/destroy', {}, {}, 200]
-]
-
-run_suite (suite)
+def suite (username):
+    return [
+        ['get root', 'get', '/', {}, '', 200],
```

Comment:
This is pretty unreadable, don't you think? Isn't it better to write these in unit test format, and have a unit test runner run them?

If you don't like the built-in `unittest`, I have good experiences with [`pytest`](https://docs.pytest.org/en/6.2.x/).

So, the `pytest` runner has built-in mechanisms for concurrent running of the tests, timing them, recording the output, etc. etc., so you don't need to build all that machinery yourself. This will improve the code readability by a lot and we'll get more flexibility out of using a well-tested tool.

Reply:
I respectfully disagree. To me, this format is eminently readable, maps very clearly to the structure of an HTTP request, and the abstraction layer is minimal (it is within the same file and fits in 100 lines). I have extensive experience writing tests in this way without relying on testing libraries which almost always involve a lot of accidental complexity.

Reply:
Chiming in here but I do agree with Rico and I think it is better to use a more traditional way of testing, with unittest. Not only is that more readable for newcomers, it will also make it easier (as Rico says) integrate with existing tooling. I am thinking also of future students or other newcomers, it is very good if they too can easily run all tests, incl. these ones.

Reply:
While I believe that the existing test runner is extremely easy to run (it's run as a mere python program), it is clear that I hold a minority view. What I suggest is that, once I work on the other suggestions, we merge this and then I open another issue where I will rewrite the e2e suite using unittest or another established tool.

Reply:
I would again recommend `pytest` over `unittest`. Trust me, you'll love it :).

Reply:
> This format is eminently readable, maps very clearly to the structure of an HTTP request,

Maybe, but wouldn't you rather map to the structure of a *scenario* you're testing, rather than HTTP semantics? Code can already do all of that!

Compare:

```python
def test_create_account_and_login():
  username, password = create_new_account()
  do_login(username, password)

  # We could write an assertion here to assert something, but the above would presumably throw if something
  # fails which would also fail the test.


# 5 concurrent workers
$ py.test -n 5 test_e2e.py
```

vs

```python
        ['login again', 'post', '/auth/login', {}, {'username': username, 'password': 'foobar2'}, 200, successfulLogin],
        ['valid signup', 'post', '/auth/signup', {}, {'username': username, 'password': 'foobar', 'email': username + '@e2e-testing.com'}, 200, successfulSignup],
```

Reply:
I understand your point but I cannot share it because I am using different criteria to judge readability and elegance. Probably a conversation to have over a beer rather than over a screen :).

Reply:
> I would again recommend `pytest` over `unittest`. Trust me, you'll love it :).

I am not against pytest but just in the interest of keeping it all the same I prefer unittest for now

Reply:
> I cannot share it because I am using different criteria to judge readability and elegance.

I understand, and it's your right to. Lord knows I've been in your exact situation a lot of times already, where something that is perfectly clear to me wasn't clear to my colleagues 😅.

I guess it's important to consciously make the trade-off when you're in this kind of situation: how important is it that "this thing is super optimized but it's hard to read and modify for other people so effectively I'm the only one who can work on it" vs "boring ugly code that anyone can read and modify but might take longer to maintain". And then decide whether the costs are worth the benefits. That trade-off will be different on a case-by-case basis...